rust/kernel/seq_file.rs

```diff
@@ -0,0 +1,213 @@
+// SPDX-License-Identifier: GPL-2.0
+
+//! Trait for defining `seq_file`s.
+//!
+//! This module allows Rust devices to implement [`struct seq_operations`] and
+//! and create a file under `/proc` based on that implementation.
+//!
+//! C header: [`include/linux/seq_file.h`](../../../include/linux/seq_file.h)
+//!
+//! Reference: <https://www.kernel.org/doc/html/latest/filesystems/seq_file.html>
+
+// Currently this module is only usable through proc_fs.
+#![cfg_attr(not(CONFIG_PROC_FS), allow(dead_code))]
+
+use core::{
+    iter::{Iterator, Peekable},
+    marker::PhantomData,
+    mem,
+    ops::{Deref, DerefMut},
+    ptr,
+};
+
+use crate::{bindings, c_types, cstr, types::PointerWrapper, CStr};
+
+#[cfg(CONFIG_PROC_FS)]
+pub use proc::proc_create_seq;
+
+/// Rust equivalent of the [`seq_operations`] interface on the C side.
+///
+/// # Example
+///
+/// ```
+/// struct Data(&'static [String]);
+///
+/// impl seq_file::SeqOperations for Data {
+///     type Item = &'static String;
+///     type Iterator = core::slice::Iter<'static, String>;
+///     type DataWrapper = Box<Self>;
+///     type IteratorWrapper = Box<Peekable<Self::Iterator>>;
+///
+///     fn start(arg: &Data) -> Option<Self::IteratorWrapper> {
+///         let iter = arg.0.iter();
+///         Box::try_new(iter.peekable()).ok()
+///     }
+///
+///     fn display(item: &Self::Item) -> &str {
+///         &item[..]
+///     }
+/// }
+/// ```
+///
+/// [`seq_operations`]: ../../../include/linux/seq_file.h
+pub trait SeqOperations {
+    /// Type produced on each iteration.
+    type Item;
+
+    /// Type created when the seq file is opened.
+    type Iterator: Iterator<Item = Self::Item>;
+
+    /// Wrapper used to store a pointer to `T` on the C side.
+    type DataWrapper: PointerWrapper + Deref<Target = Self>;
+
+    /// Wrapper used to store a pointer to the iterator on the C side.
+    type IteratorWrapper: PointerWrapper + DerefMut<Target = Peekable<Self::Iterator>>;
+
+    /// Called once each time the `seq_file` is opened.
+    fn start(arg: &Self) -> Option<Self::IteratorWrapper>;
```

Comment:
Is there a reason to return `Option<T>` instead of `Result<T>`?

Reply:
Yeah `Result` would make sense. We should probably then also convert it with `ERR_PTR` in `start_callback` then? I see that we have a conversion from kernel err pointers to Rust `Result`s, but not the other way. I'll open a separate PR to add them and then use it here.